### PR TITLE
Use futures in LQS

### LINKED ISSUE #559: Use futures::stream in LQS
Make LQS use `futures` stream to facilitate the async fetching of the receipt out of the stream.

Pre-work:
Keep #412 in mind.

Blocks #412

### PR DESCRIPTION
**Pull request by bonomat**
Thank you @LLFourn for the great support on this one.
I'm sure there is lots of stuff to cleanup now :D 

Resolves #559

### CODE REVIEW

**Comment by bonomat** [comment]
Note: doing this ticket we found an issue with pending transactions which is tracked in this ticket: #591.

File: application/ledger_query_service/src/queries/ethereum.rs [hunk 1]
```diff
@@ -73,22 +73,25 @@ impl ExpandResult for EthereumTransactionQuery {
             .collect();
 
         stream::futures_ordered(futures)
-            .filter_map(|item| item)
+            .filter_map(|item| item) // todo remove me?
```
Comment (luckysori) [inline application/ledger_query_service/src/queries/ethereum.rs]: :question: Why is this needed? :open_mouth:  Also the TODO should probably be fleshed out :stuck_out_tongue:

Comment (bonomat) [reply application/ledger_query_service/src/queries/ethereum.rs]: Good catch... it's actually needed. I've removed the TODO

File: application/ledger_query_service/src/bin/ledger_query_service.rs [hunk 2]
```diff
@@ -112,28 +131,37 @@ fn create_ethereum_routes(
...
-            transaction_processor,
-        );
-        match web3_poller {
-            Ok(listener) => listener.start(),
-            Err(e) => error!("Failed to start EthereumSimpleListener! {:?}", e),
-        }
-    });
+    let web3_client = Arc::new(Web3::new(transport));
+
+    let poller_client = Arc::clone(&web3_client);
```
Comment (LLFourn) [inline application/ledger_query_service/src/bin/ledger_query_service.rs]: 🔧 inline this clone into the `create` call below.

File: application/ledger_query_service/src/connectors/bitcoind_zmq_listener.rs [hunk 3]
```diff
@@ -1,57 +1,50 @@
...
 #[derive(DebugStub)]
-pub struct BitcoindZmqListener<P> {
-    #[debug_stub = "Context"]
-    _context: Context,
-    #[debug_stub = "Socket"]
-    socket: Socket,
-    #[debug_stub = "Processor"]
-    processor: P,
-}
+pub struct BitcoindZmqListener {}
```
Comment (LLFourn) [inline application/ledger_query_service/src/connectors/bitcoind_zmq_listener.rs]: ❌ Delete this struct since it does nothing now

File: application/ledger_query_service/src/queries/bitcoin.rs [hunk 4]
```diff
@@ -338,7 +359,20 @@ mod tests {
             confirmations_needed: 0,
         };
 
-        assert_that(&query.matches(&tx)).is_equal_to(QueryMatchResult::yes());
+        exec_future(query.matches(&tx), |result| {
+            assert_that(&result).is_equal_to(QueryMatchResult::yes());
+        });
+    }
+
+    fn exec_future<F: Fn(QueryMatchResult)>(
```
Comment (LLFourn) [inline application/ledger_query_service/src/queries/bitcoin.rs]: ❓ wouldn't just returning the result make this easier to use?

File: application/ledger_query_service/src/block_processor.rs [hunk 5]
```diff
@@ -90,49 +102,74 @@ impl<T: Transaction, B: Block<Transaction = T>, TQ: Query<T> + 'static, BQ: Quer
...
 
-        block
-            .transactions()
-            .iter()
-            .for_each(|tx| self.process_new_transaction(tx));
+        let block_futures = Self::process_new_block(self.block_queries.clone(), block.clone());
+
+        let transaction_queries = self.transaction_queries.clone();
+        let pending_transaction = self.pending_transactions.clone();
+        let txns: Vec<T> = block.transactions().iter().map(Clone::clone).collect();
```
[the remark this answers is not in the document]
Comment (bonomat) [reply application/ledger_query_service/src/block_processor.rs]: Please, feel free to do so :)
